### DISCUSSION TITLE
Is it possible to cast from `Stream Float` to `Stream Word16`?

### DISCUSSION THREAD

**Discussion by fer7una** [Copilot-Language/copilot#571] [Q&A]
Hi everyone,

I'm working on Arduino.Copilot and my goal is to calculate the PWM signal interpreted as a Word8. I have to divide the value received from analog input (Int16) by 1023, multiplying the result by 0.5:

```haskell
analogValue0 <- input a0 :: Sketch (Stream Int16) -- receive the value from A0 pin
analogValue1 <- input a1 :: Sketch (Stream Int16) -- receive the value from A1 pin
let valor = unsafeCast (abs ((analogValue0 - 511) `div` 2) :: Stream Float -- get the initial value
    analogValueF = unsafeCast analogValue1 :: Stream Float 
    factor = (analogValueF / 1023) * 0.5
    pwmValue = unsafeCast (valor * factor) :: Stream Word16 -- trying to cast from Stream Float to Stream Word16
pin3 =: pwm (unsafeCast pwmValue)
```

The problem is that there is no instance for `UnsafeCast Float Word16`. Someone who knows any alternative solution?

**Comment by ivanperez-keera** [comment]
@fdedden is this something we could add?

There's not even an instance for `UnsafeCast Float Double`, which is strange.

**Reply by ivanperez-keera** [reply]
Tagging also @RyanGlScott.

**Reply by RyanGlScott** [reply]
What is the intended semantics for casting a `Float` (e.g., `0.6`) to a `Word16`? I can think of several possible ways to handle this:

1. Reinterpret the raw bits of the `Float` as a `Word16`, assuming that the `Float` value is a [half-precision float](https://en.wikipedia.org/wiki/Half-precision_floating-point_format). This is what `cborg`'s [`floatToWord16`](https://hackage.haskell.org/package/cborg-0.2.10.0/docs/Codec-CBOR-Magic.html#v:floatToWord16) operation does, e.g.,

   ```
   λ> floatToWord16 0.6
   14541
   ```
2. Truncate the value of the `Float` to an integer, rounding towards zero. This is what `base`'s [`truncate`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:truncate) function does, and it matches the semantics of how casting a `float` to an `uint16_t` would work in C, e.g.,

   ```
   λ> truncate 0.6
   0
   ```
3. Round the `Float` value to an integer using some other rounding mode. For example, we could round `0.6` towards positive infinity:

   ```
   λ> ceiling 0.6
   1
   ```

My impression is that the semantics of `unsafeCast` tend to reflect how C performs casts, so I assume that (2) is what you want. But please double-check me here. Regardless of what choice is made here, we should be careful to document what it does.

> There's not even an instance for `UnsafeCast Float Double`, which is strange.

Indeed. There should probably be an `UnsafeCast Double Float` instance as well, since there already exist instances which lose precision outright (e.g., `UnsafeCast Int64 Int32`).

**Reply by fer7una** [reply]
I could round a `Float`, but how do I round a `Stream Float`? 
```haskell
analogValue0 <- input a0 :: Sketch (Stream Int16) -- receive the value from A0 pin
analogValue1 <- input a1 :: Sketch (Stream Int16) -- receive the value from A1 pin 
let valor = unsafeCast abs((analogValue0 - 511)`div` 2) :: Stream Float
    analogValueF = unsafeCast analogValue1 :: Stream Float
    factor = (analogValueF/1023)*0.5
    pwmValueFloat = valor*factor
    pwmValueInt = map ceiling pwmValueFloat -- couldn't match expected type [a1] with actual type Stream Float
```
What would be the semantics of doing something like this?

**Reply by RyanGlScott** [reply]
In the example above, you want to use [Copilot's version of `ceiling`](https://hackage.haskell.org/package/copilot-language-4.1/docs/Copilot-Language-Stream.html#v:ceiling), which is lifted point-wise to work over the elements of a `Stream`. (`map ceiling` doesn't make sense in your example, since there is no `map` function for `Stream`s.) For instance, if you had `stream = [0.6, 1.4, 2.6, ...]`, then `Copilot.ceiling stream = [1, 2, 3, ...]`.

Similarly, calling `unsafeCast` on a `Stream Float` value would cast each `Float` value in the `Stream` pointwise.

**Reply by ivanperez-keera** [reply]
Unless I am mistaken, @RyanGlScott is talking about how it would work. AFAICT there is no way at the moment to go from `Stream Float` to `Stream Word64` (or any other integer type). Not with `ceiling`, not with `Copilot.ceiling`, not with `unsafeCast`, not in any other way.

Please correct me if I'm mistaken. I'd be happy to be wrong about this.

**Reply by RyanGlScott** [reply]
Oops, I forgot that the type of `Copilot.stream` is `(Typed a, RealFrac a) => Stream a -> Stream a` (i.e., it doesn't go from a floating-point type to an integral type like `Prelude.ceiling`). In that case, my suggestion in https://github.com/Copilot-Language/copilot/discussions/571#discussioncomment-11560803 won't typecheck. (The statement about lifting operations pointwise still applies, though.)

**Reply by fer7una** [reply]
That's interesting, now I could round a `Stream Float` obtaining another `Stream Float`, but the problem still being how can I cast from `Stream Float` to `Stream Word8`...

**Reply by RyanGlScott** [reply]
There isn't a way to do that at present, but we would like to add such a way to Copilot. Can you comment above on which semantics listed in https://github.com/Copilot-Language/copilot/discussions/571#discussioncomment-11556202 would match your expectation on how a hypothetical `Float`-to-`Word8` cast should work?

**Reply by ivanperez-keera** [reply]
Based on the code that I saw from @fer7una (received separately), I think he's looking for something like rounding. That being said, I think _truncating would be enough_.

Also, you can implement rounding if you have truncation:
```haskell
round x
    | (x - x') >= 0.5 = x' + 1
    | otherwise       = x'
  where
    x' = truncate x
```

Haskell's prelude has an implementation that also works for negative numbers. It's a bit more convoluted. It could be extended to streams.

**Reply by fer7una** [reply]
Studying @ivanperez-keera's answer I have a provisional solution to the rounding problem, getting a `Stream Float` with rounded values and a function that receives a `Stream Float` and returns the same value in `Stream Word8`, it works by setting the list defined in the function with the proper range.

```haskell
import Copilot.Language.Stream as Copilot
import Copilot.Arduino.Uno
import Data.Foldable

main :: IO ()
main = arduino $ do
  analogValue0 <- input a0 :: Sketch (Stream Int16) -- receive the value from A0 pin
  analogValue1 <- input a1 :: Sketch (Stream Int16) -- receive the value from A1 pin 
  let valor = unsafeCast abs((analogValue0 - 511)`div` 2) :: Stream Float
      analogValueF = unsafeCast analogValue1 :: Stream Float
      factor = (analogValueF/1023)*0.5
      pwmValueFloat = valor*factor
      pwmValueFloatRound = roundStream pwmValueFloat
      pwmValueW8 = eqValue pwmValueFloatRound 
  delay =: MilliSeconds 100

-- | Round using Copilot.Language.Stream
roundStream :: Stream Float -> Stream Float
roundStream x =
  if (x - x') >= 0.5
	then x' + 1
	else x'
  where
	x' = Copilot.floor x

-- | Return a Word8 value equal to the given Float (needed to round the Float value beforehand, if not exact, the function will return 0)
eqValue :: Stream Float -> Stream Word8
eqValue target = foldl1 (.|.) (map (\ s -> mux (unsafeCast s == target) s 0) streams)
  where
	streams = map constW8 [0..255]

```

**Reply by ivanperez-keera** [reply]
Does this compile and run (generate code)?

**Reply by fer7una** [reply]
The complete code with the WheelMotor sketches compiles! But I can't test it in the simulator because it generates a program that uses 96700 bytes of program memory (maximum is 32256 bytes).

**Comment by ivanperez-keera** [comment]
I've been looking around a bit. I think, to do this efficiently, we'll have to introduce new operations.

On a related note, I think we could implement the following two classes for `Stream`s: https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Fractional and https://hackage.haskell.org/package/base-4.21.0.0/docs/Prelude.html#t:Floating.
We cannot do `RealFrac` or `RealFloat`.

**Reply by RyanGlScott** [reply]
Correct me if I'm wrong, but isn't `Stream` already [an instance of `Fractional` and `Floating`](https://github.com/Copilot-Language/copilot/blob/633c7b99ef854e704139fe24880002dfd55f7d75/copilot-language/src/Copilot/Language/Stream.hs#L86-L122)?

**Reply by ivanperez-keera** [reply]
https://i.imgur.com/ktyIhi8.gif

**Reply by ivanperez-keera** [reply]
This will sit right there in history next to when I look for the answer to something and I end up finding _my own response_ to the same question on StackOverflow #facepalm
